File: 11.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0, r = height.size()-1;
        int res = INT_MIN;
        while(l<r){
            int h = min(height[l],height[r]);
            res =max(res, h*(r-l));
            while(l<r && height[l]<=h) l++;
            while(l<r && height[r]<=h) r--;
        }
        return res;
    }
};

File: golang/11/11.go
package main

func maxArea(height []int) int {
	left, right := 0, len(height)-1
	res := -1

	for left < right {
		h := min(height[left], height[right])
		res = max(res, h*(right-left))

		for (left < right) && (height[left] <= h) {
			left++
		}

		for (left < right) && (height[right] <= h) {
			right--
		}
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
